Engine/src/ECS/Systems/BaseSystem.bf

```diff
@@ -0,0 +1,318 @@
+		/// </summary>
+		protected Queue<BaseComponent> _uninitializedComponents ~ delete _;
+
+		/// <summary>
+		/// Tracks all potentially uninitialized <see cref="SteelEngine.ECS.BaseComponent"/> objects.
+		/// All <see cref="SteelEngine.ECS.BaseComponent"/> objects will be initialized in the <see cref="Initialize"/>, <see cref="PreUpdate(GameTime)"/>, or <see cref="PostUpdate(GameTime)"/> methods.
+		/// </summary>
+		protected Type[] _registeredTypes ~ delete _;
+
+		/// <summary>
+		/// Adds a <see cref="SteelEngine.ECS.BaseComponent"/> to relevant <see cref="Components"/>. If the <see cref="SteelEngine.ECS.BaseComponent"/> is not initialized, it will be queued for initialization.
+		/// </summary>
+		/// <param name="componentToAdd"><see cref="SteelEngine.ECS.BaseComponent"/> to add.</param>
+		/// <returns>Whether or not the <see cref="SteelEngine.ECS.BaseComponent"/> was added to this BehaviorSystem's <see cref="Components"/>. Returns false if the <see cref="SteelEngine.ECS.BaseComponent"/> already existed.</returns>
+		protected virtual bool AddComponent(BaseComponent component)
+		{
+			let parent = component.Parent;
+			List<BaseComponent> entityComponents = ?;
+			if (parent == null)
+			{
+				return false;
+			}
+			bool isNewEntity = false;
+			if (!EntityToComponents.TryGetValue(parent.Id, out entityComponents))
+			{
+				isNewEntity = true;
+				entityComponents = new List<BaseComponent>();
+			}
+			for (let entityComponent in entityComponents)
+			{
+				if (component.Id == entityComponent.Id)
+				{
+					return false;
+				}
+			}
+			if (!CanBeRegistered(component))
+			{
+				if (isNewEntity)
+				{
+					delete entityComponents;
+				}
+				return false;
+			}
+
+			if (!component.IsInitialized)
+			{
+				_uninitializedComponents.Enqueue(component);
+			}
+			if (isNewEntity)
+			{
+				EntityToComponents[parent.Id] = entityComponents;
+			}
+			entityComponents.Add(component);
+			component.[Friend]IsQueuedForDeletion = false;
+			_entityRegistrationChecks.Add(parent.Id);
+			return true;
+		}
+
+		protected bool CanBeRegistered(BaseComponent component)
+		{
+			for (let type in _registeredTypes)
+			{
+				if (CanBeRegisteredAsType(component, type))
+				{
+					return true;
+				}
+			}
+			return false;
+		}
+
+		protected bool CanBeRegisteredAsType(BaseComponent component, Type type)
+		{
+			let componentType = component.GetType();
+			if (componentType.TypeId == type.TypeId || componentType.IsSubtypeOf(type))
+			{
+				return true;
+			}
+			return false;
+		}
+
+		protected void ClearEmptyEntities()
+		{
+			let entitiesToRemove = new List<EntityId>();
+			defer delete entitiesToRemove;
+
+			for (let item in EntityToComponents)
+			{
+				if (item.value == null || item.value.Count == 0)
+				{
+					entitiesToRemove.Add(item.key);
+				}
+			}
+			for (let entity in entitiesToRemove)
+			{
+				EntityToComponents.Remove(entity);

```

Comment:
Seems like the memory leak is coming from this line.

Reply:
I will look into this. Thanks!